### PR TITLE
AV-73526 | Handle rate limit error 

### PR DESCRIPTION
**Pull request by ajsqr**
## Jira

* AV-73526

## Description

Please include a summary of the fix/feature/change, including any relevant motivation and context.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

### How was this change tested and do you have evidence? _**(REQUIRED: Select at least 1)**_

- [x] Manually tested
- [ ] Unit tested
- [ ] Acceptance tested
- [ ] Unable to test / will not test (Please provide comments in section below)

### Testing

<details open>
  <summary>Testing</summary>
 Decreased the rate limit in local-dev manually to 3 and attempted creating 7 API Key resources

 Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

couchbase-capella_apikey.new_apikey5: Creating...
couchbase-capella_apikey.new_apikey2: Creating...
couchbase-capella_apikey.new_apikey1: Creating...
couchbase-capella_apikey.new_apikey4: Creating...
couchbase-capella_apikey.new_apikey6: Creating...
couchbase-capella_apikey.new_apikey3: Creating...
couchbase-capella_apikey.new_apikey: Creating...
couchbase-capella_apikey.new_apikey4: Still creating... [10s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [10s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [10s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [10s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [10s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [10s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [10s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [20s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [20s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [20s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [20s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [20s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [20s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [20s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [30s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [30s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [30s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [30s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [30s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [30s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [30s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [40s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [40s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [40s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [40s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [40s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [40s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [40s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [50s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [50s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [50s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [50s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [50s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [50s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [50s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [1m0s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [1m0s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [1m0s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [1m0s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [1m0s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [1m0s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [1m0s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [1m10s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [1m10s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [1m10s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [1m10s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [1m10s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [1m10s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [1m10s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [1m20s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [1m20s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [1m20s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [1m20s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [1m20s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [1m20s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [1m20s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [1m30s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [1m30s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [1m30s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [1m30s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [1m30s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [1m30s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [1m30s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [1m40s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [1m40s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [1m40s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [1m40s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [1m40s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [1m40s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [1m40s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [1m50s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [1m50s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [1m50s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [1m50s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [1m50s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [1m50s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [1m50s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [2m0s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [2m0s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [2m0s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [2m0s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [2m0s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [2m0s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [2m0s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [2m10s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [2m10s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [2m10s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [2m10s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [2m10s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [2m10s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [2m10s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [2m20s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [2m20s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [2m20s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [2m20s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [2m20s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [2m20s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [2m20s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [2m30s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [2m30s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [2m30s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [2m30s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [2m30s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [2m30s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [2m30s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [2m40s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [2m40s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [2m40s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [2m40s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [2m40s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [2m40s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [2m40s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [2m50s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [2m50s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [2m50s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [2m50s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [2m50s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [2m50s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [2m50s elapsed]
couchbase-capella_apikey.new_apikey4: Still creating... [3m0s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [3m0s elapsed]
couchbase-capella_apikey.new_apikey1: Still creating... [3m0s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [3m0s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [3m0s elapsed]
couchbase-capella_apikey.new_apikey3: Still creating... [3m0s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [3m0s elapsed]
couchbase-capella_apikey.new_apikey3: Creation complete after 3m0s [id=VX8Xr6llnmcTLvMkxAwQMnup5NDPBney]
couchbase-capella_apikey.new_apikey4: Creation complete after 3m0s [id=gZEbv3KMUgzUhyxZ0H5b3dHtD0y1AAtC]
couchbase-capella_apikey.new_apikey1: Creation complete after 3m1s [id=AYYCp3CKFK2PnXZX4Z157g1NFC7VQbnw]
couchbase-capella_apikey.new_apikey6: Still creating... [3m10s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [3m10s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [3m10s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [3m10s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [3m20s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [3m20s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [3m20s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [3m20s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [3m30s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [3m30s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [3m30s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [3m30s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [3m40s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [3m40s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [3m40s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [3m40s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [3m50s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [3m50s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [3m50s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [3m50s elapsed]
couchbase-capella_apikey.new_apikey: Still creating... [4m0s elapsed]
couchbase-capella_apikey.new_apikey5: Still creating... [4m0s elapsed]
couchbase-capella_apikey.new_apikey2: Still creating... [4m0s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [4m0s elapsed]
couchbase-capella_apikey.new_apikey5: Creation complete after 4m1s [id=qiOsKJosajS4nB9Duq7CkomuFD1ODxqW]
couchbase-capella_apikey.new_apikey: Creation complete after 4m1s [id=1wcE9RRDp4aES88PJdxzHZfFhkkZasBN]
couchbase-capella_apikey.new_apikey2: Creation complete after 4m1s [id=9LXaeRdBUC9bDVIq6nY9qIvFBZxxAb7P]
couchbase-capella_apikey.new_apikey6: Still creating... [4m10s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [4m20s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [4m30s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [4m40s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [4m50s elapsed]
couchbase-capella_apikey.new_apikey6: Still creating... [5m0s elapsed]
couchbase-capella_apikey.new_apikey6: Creation complete after 5m1s [id=57uxQVKCu1Dl7UwlTBR2QYTRK8kqYKAg]

Apply complete! Resources: 7 added, 0 changed, 0 destroyed.

=========================================================
Acceptance testing
=========================================================

ajayajith@Ajays-MBP-2 terraform-provider-couchbase-capella % make testacc
TF_ACC=1 go test -timeout=300m -v ./... -run TestAccOrganizationDataSource -run TestAccProjectResource -run TestAccCreateProjectWithReqFields -run TestAccCreateProjectOptFields -run TestAccValidProjectUpdate -run TestAccInvalidProjectResource -run TestAccDeleteProjectBeforeDestroy -run TestAccClusterResourceAzure -run TestAccClusterResourceGCP -run TestAccClusterResourceWithOnlyReqFieldAWS
?       github.com/couchbasecloud/terraform-provider-couchbase-capella  [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api/appservice  [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api/backup      [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api/backup_schedule     [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api/bucket      [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api/cluster     [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api/organization        [no test files]
testing: warning: no tests to run
PASS
ok      github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api     (cached) [no tests to run]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/errors  [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/provider        [no test files]
testing: warning: no tests to run
PASS
ok      github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/datasources     (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/resources       (cached) [no tests to run]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/testing [no test files]
?       github.com/couchbasecloud/terraform-provider-couchbase-capella/version  [no test files]
=== RUN   TestAccClusterResourceWithOnlyReqFieldAWS
=== PAUSE TestAccClusterResourceWithOnlyReqFieldAWS
=== CONT  TestAccClusterResourceWithOnlyReqFieldAWS
raw state map[%:16 audit.%:5 audit.created_at:2024-03-11 10:45:46.529028332 +0000 UTC audit.created_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.modified_at:2024-03-11 10:48:00.185446764 +0000 UTC audit.modified_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.version:5 availability.%:1 availability.type:multi cloud_provider.%:3 cloud_provider.cidr:10.208.250.0/23 cloud_provider.region:us-east-1 cloud_provider.type:aws configuration_type:multiNode couchbase_server.%:1 couchbase_server.version:7.2 current_state:healthy description: etag:Version: 5 id:abb708ed-d5cb-479c-9194-296ff5934eb4 name:acc_cluster_eugrmdysjc organization_id:7a99d00c-f55b-4b39-bc72-1b4cc68ba894 project_id:e14f20e3-42fb-4cb1-9928-2ac7826ffd48 service_groups.#:1 service_groups.0.%:3 service_groups.0.node.%:2 service_groups.0.node.compute.%:2 service_groups.0.node.compute.cpu:4 service_groups.0.node.compute.ram:16 service_groups.0.node.disk.%:4 service_groups.0.node.disk.iops:3000 service_groups.0.node.disk.storage:50 service_groups.0.node.disk.type:io2 service_groups.0.num_of_nodes:3 service_groups.0.services.#:3 service_groups.0.services.0:data service_groups.0.services.1:index service_groups.0.services.2:query support.%:2 support.plan:developer pro support.timezone:PT]raw state map[%:16 audit.%:5 audit.created_at:2024-03-11 10:45:46.529028332 +0000 UTC audit.created_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.modified_at:2024-03-11 10:48:00.185446764 +0000 UTC audit.modified_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.version:5 availability.%:1 availability.type:multi cloud_provider.%:3 cloud_provider.cidr:10.208.250.0/23 cloud_provider.region:us-east-1 cloud_provider.type:aws configuration_type:multiNode couchbase_server.%:1 couchbase_server.version:7.2 current_state:healthy description: etag:Version: 5 id:abb708ed-d5cb-479c-9194-296ff5934eb4 name:acc_cluster_eugrmdysjc organization_id:7a99d00c-f55b-4b39-bc72-1b4cc68ba894 project_id:e14f20e3-42fb-4cb1-9928-2ac7826ffd48 service_groups.#:1 service_groups.0.%:3 service_groups.0.node.%:2 service_groups.0.node.compute.%:2 service_groups.0.node.compute.cpu:4 service_groups.0.node.compute.ram:16 service_groups.0.node.disk.%:4 service_groups.0.node.disk.iops:3000 service_groups.0.node.disk.storage:50 service_groups.0.node.disk.type:io2 service_groups.0.num_of_nodes:3 service_groups.0.services.#:3 service_groups.0.services.0:data service_groups.0.services.1:index service_groups.0.services.2:query support.%:2 support.plan:developer pro support.timezone:PT]raw state map[%:16 audit.%:5 audit.created_at:2024-03-11 10:45:46.529028332 +0000 UTC audit.created_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.modified_at:2024-03-11 10:54:44.183857839 +0000 UTC audit.modified_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.version:11 availability.%:1 availability.type:multi cloud_provider.%:3 cloud_provider.cidr:10.208.250.0/23 cloud_provider.region:us-east-1 cloud_provider.type:aws configuration_type:multiNode couchbase_server.%:1 couchbase_server.version:7.2 current_state:healthy description:Cluster Updated. etag:Version: 11 id:abb708ed-d5cb-479c-9194-296ff5934eb4 if_match:5 name:Terraform Acceptance Test Cluster Update organization_id:7a99d00c-f55b-4b39-bc72-1b4cc68ba894 project_id:e14f20e3-42fb-4cb1-9928-2ac7826ffd48 service_groups.#:2 service_groups.0.%:3 service_groups.0.node.%:2 service_groups.0.node.compute.%:2 service_groups.0.node.compute.cpu:4 service_groups.0.node.compute.ram:16 service_groups.0.node.disk.%:4 service_groups.0.node.disk.iops:3001 service_groups.0.node.disk.storage:51 service_groups.0.node.disk.type:gp3 service_groups.0.num_of_nodes:3 service_groups.0.services.#:1 service_groups.0.services.0:data service_groups.1.%:3 service_groups.1.node.%:2 service_groups.1.node.compute.%:2 service_groups.1.node.compute.cpu:8 service_groups.1.node.compute.ram:32 service_groups.1.node.disk.%:4 service_groups.1.node.disk.iops:3001 service_groups.1.node.disk.storage:51 service_groups.1.node.disk.type:gp3 service_groups.1.num_of_nodes:2 service_groups.1.services.#:2 service_groups.1.services.0:index service_groups.1.services.1:query support.%:2 support.plan:enterprise support.timezone:IST]raw state map[%:16 audit.%:5 audit.created_at:2024-03-11 10:45:46.529028332 +0000 UTC audit.created_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.modified_at:2024-03-11 10:54:57.850897684 +0000 UTC audit.modified_by:7AsCv7QPGJ6pGQEFIau2P8qlN9OlQ6gz audit.version:16 availability.%:1 availability.type:multi cloud_provider.%:3 cloud_provider.cidr:10.208.250.0/23 cloud_provider.region:us-east-1 cloud_provider.type:aws configuration_type:multiNode couchbase_server.%:1 couchbase_server.version:7.2 current_state:healthy description:Cluster Updated. etag:Version: 16 id:abb708ed-d5cb-479c-9194-296ff5934eb4 name:Terraform Acceptance Test Cluster Update 2 organization_id:7a99d00c-f55b-4b39-bc72-1b4cc68ba894 project_id:e14f20e3-42fb-4cb1-9928-2ac7826ffd48 service_groups.#:2 service_groups.0.%:3 service_groups.0.node.%:2 service_groups.0.node.compute.%:2 service_groups.0.node.compute.cpu:4 service_groups.0.node.compute.ram:16 service_groups.0.node.disk.%:4 service_groups.0.node.disk.iops:3001 service_groups.0.node.disk.storage:51 service_groups.0.node.disk.type:gp3 service_groups.0.num_of_nodes:3 service_groups.0.services.#:1 service_groups.0.services.0:data service_groups.1.%:3 service_groups.1.node.%:2 service_groups.1.node.compute.%:2 service_groups.1.node.compute.cpu:8 service_groups.1.node.compute.ram:32 service_groups.1.node.disk.%:4 service_groups.1.node.disk.iops:3001 service_groups.1.node.disk.storage:51 service_groups.1.node.disk.type:gp3 service_groups.1.num_of_nodes:2 service_groups.1.services.#:2 service_groups.1.services.0:index service_groups.1.services.1:query support.%:2 support.plan:enterprise support.timezone:IST]--- PASS: TestAccClusterResourceWithOnlyReqFieldAWS (904.59s)
PASS
ok      github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/resources/acceptance_tests      905.042s
testing: warning: no tests to run
PASS
ok      github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/resources/acceptance_tests/security_acceptance_tests    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/schema  (cached) [no tests to run]
</details>

## Required Checklist:

- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if required)
- [x] I have run make fmt and formatted my code

## Further comments

### CODE REVIEW

File: internal/api/client.go [hunk 1]
```diff
@@ -150,38 +151,47 @@ func (c *Client) ExecuteWithRetry(
 		switch apiRes.StatusCode {
 		case endpointCfg.SuccessStatus:
 			// success case
+		case http.StatusTooManyRequests:
+			header := apiRes.Header.Get("Retry-After")
+			retryAfter, err := strconv.Atoi(header)
+			if err != nil {
+				return nil, dur, fmt.Errorf("error parsing Retry-After value from response header")
+			}
+			dur = time.Second * time.Duration(retryAfter)
```
Comment (Talina06) [inline internal/api/client.go:160]: stagger the retries with a random buffer added to this duration. not too big, ms worth...
retry limit: 100/min
API limit is per tenant

one terraform provider running, but what if the customer is creating 100 users at a time...? or such requests.

**Review by Talina06 — APPROVED** [review]

File: internal/errors/errors.go [hunk 2]
```diff
@@ -159,6 +159,10 @@ var (
 	// ErrIfMatchCannotBeSetWhileCreate is returned when if_match is set during create operation.
 	ErrIfMatchCannotBeSetWhileCreate = errors.New("if_match attribute cannot be set during create operation")
 
+	// ErrRatelimit is returned when an API key reaches the ratelimit
+	ErrRatelimit = errors.New("api key reached the ratelimit")
+
 	// ErrScopeNameMissing is returned when an expected ScopeName was not found after an import.
 	ErrScopeNameMissing = errors.New("scope Name is missing or was passed incorrectly, please check provider documentation for syntax")
+
```
Comment (Talina06) [inline internal/errors/errors.go:167]: needs to be removed

File: internal/errors/errors.go [hunk 3]
```diff
@@ -159,6 +159,10 @@ var (
 	// ErrIfMatchCannotBeSetWhileCreate is returned when if_match is set during create operation.
 	ErrIfMatchCannotBeSetWhileCreate = errors.New("if_match attribute cannot be set during create operation")
 
+	// ErrRatelimit is returned when an API key reaches the ratelimit
```
Comment (Talina06) [inline internal/errors/errors.go:162]:
```suggestion
	// ErrRatelimit is returned when the Capella API reaches a rate limit for the same API key.
```

File: internal/api/client.go [hunk 4]
```diff
@@ -196,14 +206,20 @@ func exec(ctx context.Context, fn func() (response *Response, err error), waitOn
...
 			case !goer.Is(err, errors.ErrGatewayTimeout):
 				return response, err
 			}
-			timer.Reset(waitOnReattempt)
+
+			if backOff > 0 {
+				timer.Reset(backOff)
+			} else {
+				timer.Reset(waitOnReattempt)
+			}
```
Comment (Talina06) [inline internal/api/client.go:222]:
```suggestion
			
				timer.Reset(backOff)
			
```

File: internal/api/client.go [hunk 5]
```diff
@@ -150,38 +151,47 @@ func (c *Client) ExecuteWithRetry(
...
 			Body:     responseBody,
-		}, nil
+		}, dur, nil
 	}
 
 	return exec(ctx, fn, defaultWaitAttempt)
 }
 
-func exec(ctx context.Context, fn func() (response *Response, err error), waitOnReattempt time.Duration) (*Response, error) {
+func exec(ctx context.Context, fn func() (response *Response, dur time.Duration, err error), waitOnReattempt time.Duration) (*Response, error) {
```
Comment (Talina06) [inline internal/api/client.go:189]:
```suggestion
func exec(ctx context.Context, fn func() (response *Response, dur time.Duration, err error)) (*Response, error) {
```

File: internal/api/client.go [hunk 6]
```diff
@@ -118,27 +119,27 @@ func (c *Client) ExecuteWithRetry(
 	headers map[string]string,
 ) (response *Response, err error) {
 	var requestBody []byte
+	var dur time.Duration
```
Comment (Talina06) [inline internal/api/client.go:122]:
```suggestion
	dur := defaultWaitAttempt
```

File: internal/api/client.go [hunk 7]
```diff
@@ -150,38 +151,47 @@ func (c *Client) ExecuteWithRetry(
...
 		}
 
 		return &Response{
 			Response: apiRes,
 			Body:     responseBody,
-		}, nil
+		}, dur, nil
 	}
 
 	return exec(ctx, fn, defaultWaitAttempt)
```
Comment (Talina06) [inline internal/api/client.go:186]:
```suggestion
	return exec(ctx, fn)
```

**Review by aniket-Kumar-c — APPROVED** [review]
LGTM

**Review by nidhi07kumar — APPROVED** [review]

**Review by girishmind — APPROVED** [review]
LGTM

**Comment by girishmind** [comment]
@SaicharanCB Tagging you here, so you have visibility.

**Pull request merged**